--- a/source/modules/vfs_default.c
+++ b/source/modules/vfs_default.c
@@ -862,7 +862,7 @@ static int vfswrap_ftruncate(vfs_handle_
 		SMB_BIG_UINT bsize,dfree,dsize;
 		big_len -= st.st_size;
 		big_len /= 1024; /* Len is now number of 1k blocks needed. */
-		space_avail = SMB_VFS_DISK_FREE(fsp->conn ,fsp->fsp_name,False,&bsize,&dfree,&dsize);
+		space_avail = get_dfree_info(fsp->conn ,fsp->fsp_name,False,&bsize,&dfree,&dsize);
 		if (space_avail == (SMB_BIG_UINT)-1) {
 			result = -1;
 			goto done;
