--- a/net/netfilter/nf_conntrack_proto_tcp.c
+++ b/net/netfilter/nf_conntrack_proto_tcp.c
@@ -51,7 +51,10 @@ static int nf_ct_tcp_loose __read_mostly
    will be started. */
 static int nf_ct_tcp_max_retrans __read_mostly = 3;
 
-  /* FIXME: Examine ipfilter's timeouts and conntrack transitions more
+/* Do not check the TCP window for incoming packets  */
+static int nf_ct_tcp_no_window_check __read_mostly = 1;
+
+/* FIXME: Examine ipfilter's timeouts and conntrack transitions more
      closely.  They're more complex. --RR */
 
 static const char *const tcp_conntrack_names[] = {
@@ -535,6 +538,9 @@ static bool tcp_in_window(const struct n
 	s16 receiver_offset;
 	bool res;
 
+	if (nf_ct_tcp_no_window_check)
+		return true;
+
 	/*
 	 * Get the required data from the packet.
 	 */
@@ -1512,6 +1518,13 @@ static struct ctl_table tcp_sysctl_table
 		.mode		= 0644,
 		.proc_handler	= proc_dointvec,
 	},
+	{
+		.procname       = "nf_conntrack_tcp_no_window_check",
+		.data           = &nf_ct_tcp_no_window_check,
+		.maxlen         = sizeof(unsigned int),
+		.mode           = 0644,
+		.proc_handler   = proc_dointvec,
+	},
 	{ }
 };
 
