--- a/include/linux/if_ppp.h
+++ b/include/linux/if_ppp.h
@@ -130,6 +130,8 @@ struct ifpppcstatsreq {
 	struct ppp_comp_stats stats;
 };
 
+typedef char	ppp_dev_name[IFNAMSIZ];
+
 #define ifr__name       b.ifr_ifrn.ifrn_name
 #define stats_ptr       b.ifr_ifru.ifru_data
 
@@ -166,6 +168,7 @@ struct ifpppcstatsreq {
 #define PPPIOCDISCONN	_IO('t', 57)		/* disconnect channel */
 #define PPPIOCATTCHAN	_IOW('t', 56, int)	/* attach to ppp channel */
 #define PPPIOCGCHAN	_IOR('t', 55, int)	/* get ppp channel number */
+#define	PPPIOCSREALDEV	_IOW('t', 53, ppp_dev_name) /* set real device name */
 
 #define SIOCGPPPSTATS   (SIOCDEVPRIVATE + 0)
 #define SIOCGPPPVER     (SIOCDEVPRIVATE + 1)	/* NEVER change this!! */
--- a/pppd/sys-linux.c
+++ b/pppd/sys-linux.c
@@ -648,6 +648,12 @@ static int make_ppp_unit()
 	}
 	if (x < 0)
 		error("Couldn't create new ppp unit: %m");
+        else if (devnam[0] != '\0' && if_nametoindex(devnam) > 0) {
+		x = ioctl(ppp_dev_fd, PPPIOCSREALDEV, devnam);
+		if (x < 0) {
+			error("Couldn't set ppp real device (%s): %m", devnam);
+		}
+	}
 
 	if (use_ifname[0] != 0) {
 		s = socket(PF_INET, SOCK_DGRAM, 0);
